### PR TITLE
callbacks: add a tqdm callback

### LINKED ISSUE #923: Docs could use an example for callbacks.
Hi !

Thank you for fsspec ! :heart: 🙏 

I think the docs could use an example for callbacks. I've come up with this little hook with `tqdm`, not sure if it's the right way to do it:
```python

import tqdm
import fsspec


class FsspecTqdmCallback(tqdm.tqdm):
    """Provides `update_to(n)` which uses `tqdm.update(delta_n)`."""

    def update_to(self, total_size, value, **kwargs):
        """
        total_size  : int
            Total size (in tqdm units)
        value  : int, optional
            Number of blocks transferred so far.
        """
        self.total = total_size
        return self.update(value - self.n)


storage_opts = {
    "key": KEY,
    "secret": SECRET_KEY,
}

fs = fsspec.filesystem("s3", **storage_opts)
path2distant_data = "s3://your-path"
with FsspecTqdmCallback(
    desc="test",
) as pbar:
    fs.upload(
        "local_path/",
        path2distant_data,
        recursive=True,
        callback=fsspec.callbacks.Callback(hooks={"tqdm": pbar.update_to}),
    )

```

### PR DESCRIPTION
**Pull request by Timost**
This PR adds a TqdmCallback class to display a progress bar using tqdm during fsspec operations.

This is probably wrong in many ways, please let me know what needs to be changed.

closes #923

### CODE REVIEW

File: fsspec/callbacks.py [hunk 1]
```diff
@@ -177,4 +180,42 @@ def call(self, **kwargs):
...
+            ".",
+            path2distant_data,
+            recursive=True,
+            callback=TqdmCallback(),
+        )
+    """
+
+    def __init__(self, *args, **kwargs):
+        try:
+            self._tqdm = importlib.import_module("tqdm")
```
Comment (martindurant) [inline fsspec/callbacks.py:203]:
```suggestion
            import tqdm
            self.tqdm = tqdm
```

**Comment by martindurant** [comment]
@efiop @isidentical you probably hav some typical callbacks to compare with

**Comment by dtrifiro** [comment]
@martindurant:

https://github.com/iterative/dvc/blob/5ff1d768261c759bbde01d402ada12c79758e7ce/dvc/fs/_callback.py#L6-L32

File: fsspec/callbacks.py [hunk 2]
```diff
@@ -177,4 +180,43 @@ def call(self, **kwargs):
...
+            path2distant_data,
+            recursive=True,
+            callback=TqdmCallback(),
+        )
+    """
+
+    def __init__(self, *args, **kwargs):
+        try:
+            import tqdm
+            self.tqdm = tqdm
```
Comment (martindurant) [inline fsspec/callbacks.py:201]:
```suggestion
            self._tqdm = tqdm
```

**Comment by martindurant** [comment]
I think it's nice to have, actually - code that lives in the code, and is immediately useful to anyone calling it.

File: fsspec/callbacks.py [hunk 3]
```diff
@@ -177,4 +177,43 @@ def call(self, **kwargs):
...
+            path2distant_data,
+            recursive=True,
+            callback=TqdmCallback(),
+        )
+    """
+
+    def __init__(self, *args, **kwargs):
+        try:
+            import tqdm
+            self._tqdm = tqdm
```
Comment (martindurant) [inline fsspec/callbacks.py:201]:
```suggestion
            import tqdm

            self._tqdm = tqdm
```

File: fsspec/callbacks.py [hunk 4]
```diff
@@ -177,4 +177,44 @@ def call(self, **kwargs):
...
+
+            self._tqdm = tqdm
+        except ImportError as exce:
+            raise ImportError(
+                "Using TqdmCallback requires tqdm to be installed"
+            ) from exce
+        super().__init__(*args, **kwargs)
+
+    def set_size(self, size):
+        self.tqdm = self._tqdm.tqdm(desc="test", total=size)
```
Comment (Timost) [inline fsspec/callbacks.py:210]: Maybe we could the tqdm `desc` parameter somewhere ?

**Comment by Timost** [comment]
Hi @martindurant thank you for doing the clean up !
🙏 I was offline the last few days.

I have a small question regarding the `desc=` parameter of tqdm.

File: fsspec/callbacks.py (hunk 4 above)
Comment (martindurant) [reply fsspec/callbacks.py:210]: Oh, yes please do add this

Comment (Timost) [reply fsspec/callbacks.py:210]: 👍Here it is: https://github.com/fsspec/filesystem_spec/pull/932

**Pull request merged**